pkg/operator/ceph/spec/label.go

```diff
@@ -38,10 +37,18 @@ const (
 func addCephVersionLabel(cephVersion version.CephVersion, labels map[string]string) {
 	// cephVersion.String() returns a string with a space in it, and labels in k8s are limited to
 	// alphanum characters plus '-', '_', '.'
-	labels[CephVersionLabelKey] = fmt.Sprintf("%d.%d.%d-%d",
+	labels[CephVersionLabelKey] = GetCephVersionLabel(cephVersion)
+}
+
+func GetCephVersionLabel(cephVersion version.CephVersion) string {
+	return fmt.Sprintf("%d.%d.%d-%d",
 		cephVersion.Major, cephVersion.Minor, cephVersion.Extra, cephVersion.Build)
 }
 
+func ExtractCephVersionFromLabel(labelVersion string) (*version.CephVersion, error) {
+	return version.ExtractCephVersion(fmt.Sprintf("ceph version %s", labelVersion))
```

Comment:
Here we are in a way relying on implementation detail of the `version.ExtractCephVersion` method. This isn't wrong, but I think it could be worthwhile to separate these concerns more. This is also related to the `fmt.Sprintf("%d.%d.%d-%d"` above.

What if, instead, we made a little more general method in `version` that would give the 4-point version string given a `version.CephVersion` and another method that would give the `version.CephVersion` given a 4-point version string? We could use those methods in the `GetCephVersionLabel` and `ExtractCephVersionFromLabel` methods here. It would use a little more code, but I think we'd have more room to suport things in the future in case `ceph --version` starts to print something other than `"ceph version %s"`.

What are your thoughts?

P.S.: very glad to have `GetCephVersionLabel` and `ExtractCephVersionFromLabel` methods. I think this is a good call.

Reply:
I actually implemented this suggestion initially, but to my mind it was a bit more cumbersome than useful. The thing that tipped me fully over was the need to feed both a build-version and simple format in to version.go in order to usefully consume its parsing (which I didn't want to fully replicate). At that point it felt like we were reliant enough on the version.go implementation anyway that it wasn't a big win.

The fact that we still feed ExtractCephVersionFromLabel through the spec module in label.go  gives this module authority over its formatting of the version. I used this simpler implementation now because we can; if the underlying output of ceph --version changes, at least we have a place to change it now.

Still, I can absolutely see the alternate decision that you suggest (I did initially implement it, after all). If we go this way it's simple enough to put back in. @leseb, thoughts?

Reply:
The current implementation doesn't bug me that much, I think it's fine for now.